File: 2018/7.rb
require 'set'
def instantiate
	lines = File.open("7.txt").readlines.map(&:strip)
	prereqs = Hash.new { |prereqs, key| prereqs[key] =  []}
	postreqs = Hash.new { |postreqs, key| postreqs[key] =  []}
	actions = Set[]
	lines.each do |line|
		m = line.match /Step (.) must be finished before step (.) can begin./
		prereqs[m[1]] << m[2]
		postreqs[m[2]] << m[1]
		actions << m[1] << m[2]
	end
	q = []
	actions.each {|action| q << action if !postreqs.keys.include? action }; q.sort!
	return prereqs, postreqs, q
end

def p1
	prereqs, postreqs, q = instantiate
	result = ""
	while !q.empty?
		action = q.shift
		prereqs[action].each do |postreq|
			postreqs[postreq] -= [action]
			q << postreq if postreqs[postreq].empty?
		end
		result << action
		q.sort!
	end
	return result
end

def timeCalc(action)
	return 60 + "ABCDEFGHIJKLMNOPQRSTUVWXYZ".index(action) + 1
end

def p2
	prereqs, postreqs, q = instantiate
	result = ""
	workers = {0 => "", 1 => "", 2 => "", 3 => "", 4 => ""}
	actions = {}
	time = -1
	while !q.empty? or workers.values.join != ""
		workers.keys.each do |worker|
			if actions[workers[worker]] == 1
				prereqs[workers[worker]].each do |postreq|
					postreqs[postreq] -= [workers[worker]]
					q << postreq if postreqs[postreq].empty?
				end
				q.sort!
				workers[worker] = ""
			else
				actions[workers[worker]] -= 1 if workers[worker] != ""
			end
			if workers[worker] == "" and !q.empty?
				workers[worker] = q.shift 
				actions[workers[worker]] = timeCalc(workers[worker])
			end
		end
		time += 1
	end
	return time
end

=begin part 1
	This actually came pretty quickly after I attempted it a second time (after about 10 days time)
	the double prereq/postreq dictionaries was very useful and useful as a double-edged tree type structure.
	I don't know if this method really counts as a toposort, but that's obviously what this was.
=end
puts p1

=begin part 2
	Man this also took me a lot longer than it should have. I figured out the structure pretty quickly but just
	couldn't get it right -- lots of bugs :(
=end
puts p2

File: 2018/7.py
class Node():
	def __init__(self, step, prereqs):
		self.prereqs = prereqs
		self.step = step
	def addPrereq(self, prereq):
		self.prereqs.append(prereq)
	def removePrereq(self, prereq):
		self.prereqs.remove(prereq)

class Worker():
	def __init__(self):
		self.busy = False
		self.project = ''
		self.timeLeft = -1


def addToAvailable(nodes, available):
	i = 0
	while i < len(nodes):
		if len(nodes[i].prereqs) == 0:
			available.append(nodes[i].step)
			nodes.pop(i)
		else:
			i += 1
	return nodes, sorted(available)

#Lookup Table
alpha = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
times = {}
for char in alpha:
	times[char] = 60 + alpha.index(char) + 1


with open("day7_input.txt", 'r') as infile:
	directions = infile.readlines()
nodes = []
for direction in directions:
	prereq = direction[direction.index(" ")+1]
	step = direction[direction.index("can")-2]
	exists = True
	for node in nodes:
		if node.step == step:
			node.addPrereq(prereq)
			exists = False
	if exists:
		nodes.append(Node(step, [prereq]))
	exists = True
	for node in nodes:
		if node.step == prereq:
			exists = False
	if exists:
		nodes.append(Node(prereq, []))

nodes, available = addToAvailable(nodes, [])

done = []
#Part 1
'''
while len(available) > 0:
	done.append(available.pop(0))
	for node in nodes:
		for finishedStep in done:
			if finishedStep in node.prereqs:
				node.removePrereq(finishedStep)
	nodes, available = addToAvailable(nodes, available)

print(''.join(done))
'''

#Part 2

workers = []
for i in range (5):
	workers.append(Worker())
time = 0
workersBusy = 0
while len(nodes) > 0 or workersBusy > 0:
	for worker in workers:
		if worker.timeLeft == 0:
			worker.busy = False
			done.append(worker.project)
			worker.timeLeft -= 1
			workersBusy -= 1
		if not worker.busy:
			if len(available) > 0:
				worker.busy = True
				worker.project = available.pop(0)
				worker.timeLeft = times[worker.project]
				workersBusy += 1
		else:
			worker.timeLeft -= 1
	for node in nodes:
		for finishedStep in done:
			if finishedStep in node.prereqs:
				node.removePrereq(finishedStep)
	nodes, available = addToAvailable(nodes, available)
	time += 1
print (time)
